README
** LibRaw-demosaic-pack-GPL3 **


THIS IS RawTherapee tracking branch. Do not use it for anything, this branch is used to hold unmodified files from RawTherapee source tree


Changelog
2011-01-13 Alex Tutubalin <lexa@lexa.ru>
	* removed exposure correction code due to instability

2011-01-10 Alex Tutubalin <lexa@lexa.ru>
	* OpenMP for green equilibration
	* Expocorrection before interpolation by Jackues Desmis

2010-12-20 Alex Tutubalin <lexa@lexa.ru>
        * Green equilibration, CFA pixel cleaning and CFA line denoise by
	  Emil Martinec. Code adaped to LibRaw by Jacques Desmis

2010-12-13 Alex Tutubalin <lexa@lexa.ru>
	* Imported new AMaZE and CA_correct code provided by Jacques Desmis

2010-11-27 Alex Tutubalin <lexa@lexa.ru>
	* malloc changed to calloc to make valgrind happy with
	  uninitialized values

2010-11-20 Alex Tutubalin <lexa@lexa.ru>
	* Imported actual AMaZE implementation from RawTherapee
	* OpenMP with LIBRAW_USE_OPENMP
	   
2010-11-15 Alex Tutubalin <lexa@lexa.ru>
	* Initial revision


green_equi.c
// CFA pixel cleaning via directional average
// by Emil Martinec
// 2/18/2010
#define TS 256	 // Tile size
//modification OMP J.Desmis - december 2010
//#define CLASS
/*#define ushort UshORt
typedef unsigned char uchar;
 typedef unsigned short ushort;*/

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <float.h>
#include <limits.h>
#include <math.h>
#include <setjmp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>


#define SQR(x) ((x)*(x))


void CLASS green_equilibrate(float thresh)//for dcraw implementation
{  
	// local variables
	static const int border=8;
	static const int border2=16;
	static const int v1=TS, v2=2*TS, v3=3*TS, /*v4=4*TS,*/ p1=-TS+1, p2=-2*TS+2, p3=-3*TS+3, m1=TS+1, m2=2*TS+2, m3=3*TS+3;
	
//+	int height=H, width=W; //for RT only
	int top, left; 

	int verbose=1;
	
	static const float eps=1.0;	//tolerance to avoid dividing by zero
	//static const float thresh=0.03;	//threshold for performing green equilibration; max percentage difference of G1 vs G2  
	// G1-G2 differences larger than this will be assumed to be Nyquist texture, and left untouched
	static const float diffthresh=0.25; //threshold for texture, not to be equilibrated

	double dt;
	clock_t t1, t2;
	
	//clock_t t1_main,       t2_main       = 0;
	
	// start
#ifdef DCRAW_VERBOSE
	if (verbose) fprintf(stderr,_("Green equilibration v1 OMP [E.Martinec] %1.3f...\n"),thresh);	
#endif
	
	
	
	t1 = clock();
#if defined (LIBRAW_USE_OPENMP)	
#pragma omp parallel
#endif
{		
	int top,left;
			char		*buffer;			// TS*TS*16
			float         (*cfa);		// TS*TS*4
			float         (*checker);			// TS*TS*4
			float         (*gvar);			// TS*TS*4
			float         (*gdiffv);			// TS*TS*4
			float         (*gdiffh);			// TS*TS*4
			
			/* assign working space */
			buffer = (char *) calloc((5*sizeof(float)+sizeof(int))*TS*TS,1);
			//merror(buffer,"green_equil()");
			memset(buffer,0,5*sizeof(float)*TS*TS);
			
			cfa         = (float (*))		buffer;
			checker		= (float (*))			(buffer +	sizeof(float)*TS*TS);
			gvar		= (float (*))			(buffer +	2*sizeof(float)*TS*TS);
			gdiffv		= (float (*))			(buffer +	3*sizeof(float)*TS*TS);
			gdiffh		= (float (*))			(buffer +	4*sizeof(float)*TS*TS);
	

	
	// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	
	// Fill G interpolated values with border interpolation and input values
	// Main algorithm: Tile loop
//#pragma omp parallel for shared(image,height,width) private(top,left) schedule(dynamic)
#if defined (LIBRAW_USE_OPENMP)
#pragma omp for schedule(dynamic) nowait 
#endif	
	for (top=0; top < height-border; top += TS-border2)
		for (left=0; left < width-border; left += TS-border2) {
			int bottom = MIN( top+TS,height);
			int right  = MIN(left+TS, width);
			int numrows = bottom - top;
			int numcols = right - left;
			
			int row, col;
			int rr, cc, c, indx;
			int vote1, vote2;
			
			float val1;
			
			float gin, gse, gsw, gne, gnw, wtse, wtsw, wtne, wtnw;
			float gu, gd, gl, gr;
			float mcorr, pcorr;
			float ginterp;
			float diffvarh, diffvarv, hvwt;
			// rgb from input CFA data
			/* rgb values should be floating point number between 0 and 1 
			 after white balance multipliers are applied */
			for (rr=0; rr < numrows; rr++)
				for (row=rr+top, cc=0; cc < numcols; cc++) {
					col = cc+left;
					cfa[rr*TS+cc] = image[row*width+col][FC(row,col)];//for dcraw implementation
//					cfa[rr*TS+cc] = ri->data[row][col]; 
				}
			
			//The green equilibration algorithm starts here
			
			for (rr=2; rr < numrows-2; rr++)
				//for (cc=3-(FC(rr,2)&1), indx=rr*TS+cc; cc < numcols-2; cc+=2, indx+=2) {
				for (indx=rr*TS+2; indx < rr*TS+numcols-2; indx++) {
					
					if (FC(rr,indx)&1) {
						pcorr = (cfa[indx+p1]-cfa[indx])*(cfa[indx-p1]-cfa[indx]);
						mcorr = (cfa[indx+m1]-cfa[indx])*(cfa[indx-m1]-cfa[indx]);
						
						if (pcorr>0 && mcorr>0) {checker[indx]=1;} else {checker[indx]=0;}
						
						//checker[indx]=1;//test what happens if we always interpolate
					} else {
						gu=cfa[indx-v1]+0.5*(cfa[indx]-cfa[indx-v2]);
						gd=cfa[indx+v1]+0.5*(cfa[indx]-cfa[indx+v2]);
						gl=cfa[indx-1]+0.5*(cfa[indx]-cfa[indx-2]);
						gr=cfa[indx+1]+0.5*(cfa[indx]-cfa[indx+2]);
						
						gdiffh[indx] = SQR((gl-gr)/(eps+gl+gr));
						gdiffv[indx] = SQR((gu-gd)/(eps+gu+gd));
						
						//gvar[indx] = 0.25*(gu*gu+gd*gd+gl*gl+gr*gr)-SQR(0.25*(gu+gd+gl+gr));
					}
				}
						

			
			//now smooth the cfa data
			for (rr=6; rr < numrows-6; rr++)
				for (cc=7-(FC(rr,2)&1), indx=rr*TS+cc; cc < numcols-6; cc+=2, indx+=2) {
					if (checker[indx]) {
						
						diffvarh = eps+(gdiffh[indx-v1]+gdiffh[indx-1]+gdiffh[indx+1]+gdiffh[indx+v1]);
						diffvarv = eps+(gdiffv[indx-v1]+gdiffv[indx-1]+gdiffv[indx+1]+gdiffv[indx+v1]);
						hvwt = fabs(diffvarv-diffvarh)/(diffvarv+diffvarh);
						
						
						vote1=(checker[indx-v2]+checker[indx-2]+checker[indx+2]+checker[indx+v2]);
						vote2=(checker[indx-m1]+checker[indx+p1]+checker[indx-p1]+checker[indx+m1]);
						if (vote1>0 && vote2>0 && hvwt<diffthresh) {
							//pixel interpolation
							
							gin=cfa[indx];
							
							gse=(cfa[indx+m1])+0.5*(cfa[indx]-cfa[indx+m2]);
							gnw=(cfa[indx-m1])+0.5*(cfa[indx]-cfa[indx-m2]);
							gne=(cfa[indx+p1])+0.5*(cfa[indx]-cfa[indx+p2]);
							gsw=(cfa[indx-p1])+0.5*(cfa[indx]-cfa[indx-p2]);
							
							
							
							wtse=1/(eps+SQR(cfa[indx+m2]-cfa[indx])+SQR(cfa[indx+m3]-cfa[indx+m1]));
							wtnw=1/(eps+SQR(cfa[indx-m2]-cfa[indx])+SQR(cfa[indx-m3]-cfa[indx-m1]));
							wtne=1/(eps+SQR(cfa[indx+p2]-cfa[indx])+SQR(cfa[indx+p3]-cfa[indx+p1]));
							wtsw=1/(eps+SQR(cfa[indx-p2]-cfa[indx])+SQR(cfa[indx-p3]-cfa[indx-p1]));
							
							ginterp=(gse*wtse+gnw*wtnw+gne*wtne+gsw*wtsw)/(wtse+wtnw+wtne+wtsw);
							
							if (/*(SQR(ginterp-gin) > 0.125*(gvar[indx-1]+gvar[indx+1]+gvar[indx-v1]+gvar[indx+v1])) &&*/ ((ginterp-gin) < thresh*(ginterp+gin)) ) {
								cfa[indx]=0.5*(ginterp+gin);
							}
							
						}
					}
				}
			
			// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
			
			// copy smoothed results back to image matrix
			for (rr=border; rr < numrows-border; rr++)
				for (row=rr+top, cc=border+1-(FC(rr,2)&1), indx=rr*TS+cc; cc < numcols-border; cc+=2, indx+=2) {
					if (cfa[indx]<1) continue;
					col = cc + left;
					c = FC(row,col);
					image[row*width + col][c] = CLIP((int)(cfa[indx] + 0.5)); //for dcraw implementation
					//ri->data[row][col] = CLIP((int)(cfa[indx] + 0.5)); 
				} 

			// clean up
		//}
		
		}
		free(buffer);

	
	//done
	}
	t2 = clock();
	dt = ((double)(t2-t1)) / CLOCKS_PER_SEC;
#ifdef DCRAW_VERBOSE
	
	if (verbose) {
		fprintf(stderr,_("elapsed time = %5.3fs\n"),dt);
				}
#endif				
	
}
#undef TS


expo_before.c
//exposure correction before interpolation
// use FFD interpolation (provisional) from Luis Sanz (complete by J.Desmis for OMP) ==> calculate luminosity
// inspired from work Guillermo Luijk and Manuel LLorens(Perfectraw)
//Jacques Desmis 31 december 2010
// I use OMP
  // This function uses parameters:
  //      exposure (lineal): 2^(-8..0..8): currently 0.5 +3
  //      preserve (log)   : 0..8 : currently 0.1 1
//OMP for Libraw


void CLASS exp_bef(float expos, float preser)
{
	double dt,dT2;
	clock_t t1, t2,t3,t4;
	float Yp, exposure2, K, EV;
	float LUT[65536];
	int i;
	int row,col;
	ushort (*img)[4];// to save configuration (provisional interpolation)
#ifdef DCRAW_VERBOSE	
	if (verbose) fprintf (stderr,_("exposure before interpolation...[J.Desmis]\n"));
#endif	
	img = (ushort (*)[4]) calloc (height*width, sizeof *img);// to save interpolation
	memcpy(img,image,height*width*sizeof *img);//save configuration
	// I use FDD interpolate from Luis Sanz , very fast and good, one can change for another
	border_interpolate(3);
	t1 = clock();
#if defined (LIBRAW_USE_OPENMP)	
#pragma omp parallel  
#endif
{
int indx,c,f[2],g[2],d;
int u=width,v=2*u,w=3*u;
#if defined (LIBRAW_USE_OPENMP)	
#pragma omp for schedule(dynamic) nowait
#endif
	for (int row=3; row < height-3; row++)
		for (int col=3+(FC(row,1)&1),c=FC(row,col),indx=row*width+col; col<u-3; col+=2,indx+=2)
			if ((abs(image[indx-2][c]-image[indx][c]-image[indx-1][1]+image[indx+1][1])+abs(image[indx][c]-image[indx+2][c]-image[indx-1][1]+image[indx+1][1])+abs(image[indx][c]-image[indx-2][c]-image[indx-1][1]+image[indx-3][1])+abs(image[indx][c]-image[indx+2][c]-image[indx+1][1]+image[indx+3][1]))<(abs(image[indx-v][c]-image[indx][c]-image[indx-u][1]+image[indx+u][1])+abs(image[indx][c]-image[indx+v][c]-image[indx-u][1]+image[indx+u][1])+abs(image[indx][c]-image[indx-v][c]-image[indx-u][1]+image[indx-w][1])+abs(image[indx][c]-image[indx+v][c]-image[indx+u][1]+image[indx+w][1])))
				image[indx][1]=ULIM((image[indx-1][1]+image[indx+1][1]+(2*image[indx][c]-image[indx-2][c]-image[indx+2][c])/2)>>1,image[indx-1][1],image[indx+1][1]);
			else
				image[indx][1]=ULIM((image[indx-u][1]+image[indx+u][1]+(2*image[indx][c]-image[indx-v][c]-image[indx+v][c])/2)>>1,image[indx-u][1],image[indx+u][1]);
#if defined (LIBRAW_USE_OPENMP)
#pragma omp for schedule(dynamic) nowait
#endif
	for (int row=1; row < height-1; row++)
		for (int col=1+(FC(row,2)&1),c=FC(row,col+1),d=2-c,indx=row*width+col; col<u-1; col+=2,indx+=2) {
			image[indx][c]=CLIP((image[indx-1][c]+image[indx+1][c]+2*image[indx][1]-image[indx-1][1]-image[indx+1][1])>>1);
			image[indx][d]=CLIP((image[indx-u][d]+image[indx+u][d]+2*image[indx][1]-image[indx-u][1]-image[indx+u][1])>>1);
		}
#if defined (LIBRAW_USE_OPENMP)
#pragma omp for schedule(dynamic) nowait
#endif
	for (int row=1; row < height-1; row++)
		for (int col=1+(FC(row,1)&1),c=2-FC(row,col),indx=row*width+col; col<u-1; col+=2,indx+=2) {
			f[0]=abs(image[indx-1-u][c]-image[indx+u+1][c])+abs(image[indx-1-u][1]-image[indx][1])+abs(image[indx+u+1][1]-image[indx][1]);
			f[1]=abs(image[indx+1-u][c]-image[indx+u-1][c])+abs(image[indx+1-u][1]-image[indx][1])+abs(image[indx+u-1][1]-image[indx][1]);
			g[0]=image[indx-1-u][c]+image[indx+u+1][c]+2*image[indx][1]-image[indx-1-u][1]-image[indx+u+1][1];
			g[1]=image[indx+1-u][c]+image[indx+u-1][c]+2*image[indx][1]-image[indx+1-u][1]-image[indx+u-1][1];
			image[indx][c]=CLIP(g[f[0]>=f[1]]>>1);
		}
		}
	t2 = clock();
// calculate CIE luminosity
float *YY;
YY = (float *)calloc(width*height,sizeof *YY);// for CIE luminosity

#pragma omp parallel default(shared)  
{
#if defined (LIBRAW_USE_OPENMP)	
#pragma omp for  
#endif
	 for(int i=0;i<height*width;i++){
           YY[i]=CLIP(0.299*(float)image[i][0]+0.587*(float)image[i][1]+0.114*(float)image[i][2]); // CIE luminosity
				}
}				
	memcpy(image,img,width*height*sizeof *img); // restore configuration
	free(img);	  //free memory
	//exposure correction inspired from G.Luijk
 if(preser==0.0){	// protect highlights 
#ifdef DCRAW_VERBOSE
  if(verbose)  fprintf (stderr,_("without highlight preservation\n"));
#endif  
#if defined (LIBRAW_USE_OPENMP)	  
#pragma omp parallel for  shared(expos)
#endif
	    for(int i=0;i<height*width;i++){
		for(int c=0;c<4;c++) image[i][c]=CLIP((float)image[i][c]*expos);}
  }else{
    // Exposure correction with highlight preservation
#ifdef DCRAW_VERBOSE	
  if(verbose)  fprintf (stderr,_("with highlight preservation\n"));
#endif  
    if(expos>1){
      K=65535/expos*exp(-preser*log((double) 2));
      for(int j=0;j<=65535;j++) LUT[(int)j]=CLIP(((65535-K*expos)/(65535-K)*(j-65535)+65535)/j);

#if defined (LIBRAW_USE_OPENMP)		  
#pragma omp parallel for  shared(expos)
#endif
      for(int i=0;i<height*width;i++){
        if(YY[i]<K){
          for(int c=0;c<4;c++) image[i][c]=CLIP((float)image[i][c]*expos); 
        }else{
          exposure2=LUT[(int)YY[i]];
          for(int c=0;c<4;c++)  image[i][c]=CLIP((float)image[i][c]*exposure2);
        }
      }
    }
	else{
      float EV=log(expos)/log(2.0);                              // Convert exp. linear to EV
      float K=65535.0*exp(-preser*log((double) 2));
      for(int j=0;j<=65535;j++) LUT[(int)j]=CLIP(exp(EV*(65535.0-j)/(65535.0-K)*log((double) 2)));
#if defined (LIBRAW_USE_OPENMP)		  
#pragma omp parallel for  shared(expos) 
#endif  
      for(int i=0;i<height*width;i++){
        if(YY[i]<K){
          for(int c=0;c<4;c++)  image[i][c]=CLIP((float)image[i][c]*expos);
        }else{
          float exposure2=LUT[(int)YY[i]];
          for(int c=0;c<4;c++)  image[i][c]=CLIP((float)image[i][c]*exposure2);
        }
      }
    }	
}
free(YY);

	t4=clock();	
		dT2 = ((double)(t4 - t1)) / CLOCKS_PER_SEC;
#ifdef DCRAW_VERBOSE		
		if (verbose) fprintf (stderr,_("\t exposure before interpolation done in %5.3fs.\n"),dT2);
#endif		
		
}

cfa_linedn_new.c
// CFA line denoise by DCT filtering
// copyright Emil Martinec
// all rights reserved
// 3/12/2010 and november 2010
#define TS 256		// Tile size
//modified by Jacques Desmis - december 2010
//#define CLASS


#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <float.h>
#include <limits.h>
#include <math.h>
#include <setjmp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "shrtdct_float.c"


#define SQRF(x) (((float)x)*((float)x))
#define ABS(x) (((int)(x) ^ ((int)(x) >> 31)) - ((int)(x) >> 31))
#define MIN(a,b) ((a) < (b) ? (a) : (b))
#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define LIM(x,min,max) MAX(min,MIN(x,max))
#define ULIM(x,y,z) ((y) < (z) ? LIM(x,y,z) : LIM(x,z,y))
#define CLIP(x) LIM(x,0,65535)

void CLASS cfa_linedn(float linenoise){  
	double dt;
	clock_t t1, t2;

	float eps=1e-5;			//tolerance to avoid dividing by zero
	const float clip_pt = (1.0 / MIN(MIN(pre_mul[0],pre_mul[1]),pre_mul[2]));  // for highlights
	const float gauss[5] = {0.20416368871516755, 0.18017382291138087, 0.1238315368057753, 0.0662822452863612, 0.02763055063889883};
	const float rolloff[8] = {0, 0.135335, 0.249352, 0.411112, 0.606531, 0.800737, 0.945959, 1}; //gaussian with sigma=3
	const float window[8] = {0, .25, .75, 1, 1, .75, .25, 0}; //sine squared
	float aarr[4][8][8], *bbrr[4][8], **dctblock[4];
	for (int j=0; j<4; j++) {
		for (int i = 0; i < 8; i++) 
			bbrr[j][i] = aarr[j][i];
		dctblock[j] = bbrr[j];
	}
	if (verbose) fprintf (stderr,_("CFA line denoise v2b OMP [E.Martinec] %1.4f \n"), linenoise);
	t1 = clock();
	
	float noisevar=SQRF(3*linenoise); // _noise_ (as a fraction of saturation) is input to the algorithm
#pragma omp parallel
{	
	char			*buffer;			// TS*TS*16
	float        	(*cfain);			// TS*TS*4
	float         	(*cfablur);		// TS*TS*4
	float         	(*cfadiff);		// TS*TS*4
	float         	(*cfadn);			// TS*TS*4
	
	

	
	// assign working space
	buffer = (char *) calloc((3*sizeof(float)+sizeof(int))*TS*TS,1);// 16
	cfain			= (float (*))		buffer; //pointers to rows of array
 	cfablur			= (float (*))			(buffer +  sizeof(float)*TS*TS); //4
	cfadiff			= (float (*))			(buffer +  2*sizeof(float)*TS*TS);//8
	cfadn			= (float (*))			(buffer +  3*sizeof(float)*TS*TS);//12
	
	
	// Main algorithm: Tile loop
#if defined (LIBRAW_USE_OPENMP)
#pragma omp for schedule(dynamic) nowait 
#endif	
	for (int top=0; top < height-16; top += TS-32)
		for (int left=0; left < width-16; left += TS-32) {
			int bottom = MIN( top+TS,height);
			int right  = MIN(left+TS, width);
			int rr1 = bottom - top;
			int cc1 = right - left;
 			int numrows = bottom - top;
 			int numcols = right - left;
			// load CFA data
			/* rgb values should be floating point number between 0 and 1 
			 after white balance multipliers are applied;
			 data should be in linear gamma space */
			for (int rr=0; rr < rr1; rr++)
				for (int cc=0, row=rr+top; cc < cc1; cc++) {
					int col = cc+left;
					int c = FC(rr,cc);
					cfain[(rr)*TS+cc] = image[row*width+col][c]/65535.0; 
				}
			//pad the block to a multiple of 16 on both sides
			if (cc1 < TS) {
				int indx=cc1 % 16;
				for (int i=0; i<(16-indx); i++) 
					for (int rr=0; rr<rr1; rr++) 
						cfain[(rr)*TS+cc1+i+1]=cfain[(rr)*TS+cc1-i];
				cc1 += 16-indx;
			}
			if (rr1 < TS) {
				int indx=rr1 % 16;
				for (int i=0; i<(16-indx); i++)
					for (int cc=0; cc<cc1; cc++)
						cfain[(rr1+i+1)*TS+cc]=cfain[(rr1-i)*TS+cc];
				rr1 += 16-indx;
			}
			//The cleaning algorithm starts here
			//gaussian blur of CFA data
			for (int rr=8; rr < rr1-8; rr++)
				for (int cc=0; cc < cc1; cc++) {
					cfablur[(rr)*TS+cc]=gauss[0]*cfain[(rr)*TS+cc];
					for (int i=1; i<5; i++) {
						cfablur[(rr)*TS+cc] += gauss[i]*(cfain[(rr-2*i)*TS+cc]+cfain[(rr+2*i)*TS+cc]);
					}
				}
			for (int rr=8; rr < rr1-8; rr++)
				for (int cc=8; cc < cc1-8; cc++) {
					cfadn[(rr)*TS+cc] = gauss[0]*cfablur[(rr)*TS+cc];
					for (int i=1; i<5; i++) {
						cfadn[(rr)*TS+cc] += gauss[i]*(cfablur[(rr)*TS+cc-2*i]+cfablur[(rr)*TS+cc+2*i]);
					}
					cfadiff[(rr)*TS+cc]=cfain[(rr)*TS+cc]-cfadn[(rr)*TS+cc]; // hipass cfa data
				}
			//begin block DCT
			float linehvar[4], linevvar[4], noisefactor[4][8][2], coeffsq;
			#pragma omp critical  // exclude this part of code from OMP
	{
			for (int rr=8; rr < numrows-22; rr+=8) // (rr,cc) shift by 8 to overlap blocks
				for (int cc=8; cc < numcols-22; cc+=8) {
					for (int ey=0; ey<2; ey++) // (ex,ey) specify RGGB subarray
						for (int ex=0; ex<2; ex++) {
							//grab an 8x8 block of a given RGGB channel
							for (int i=0; i<8; i++) 
								for (int j=0; j<8; j++) {
									dctblock[2*ey+ex][i][j]=cfadiff[(rr+2*i+ey)*TS+cc+2*j+ex];
								}
							ddct8x8s(-1, dctblock[2*ey+ex]); //forward DCT
						}
					for (int ey=0; ey<2; ey++) // (ex,ey) specify RGGB subarray
						for (int ex=0; ex<2; ex++) {
							linehvar[2*ey+ex]=linevvar[2*ey+ex]=0;
							for (int i=4; i<8; i++) {
								linehvar[2*ey+ex] += SQRF(dctblock[2*ey+ex][0][i]);
							linevvar[2*ey+ex] += SQRF(dctblock[2*ey+ex][i][0]);		
							}
							//Wiener filter for line denoising; roll off low frequencies
							for (int i=1; i<8; i++) {
								coeffsq = SQRF(dctblock[2*ey+ex][i][0]);//vertical
								noisefactor[2*ey+ex][i][0] = coeffsq/(coeffsq+rolloff[i]*noisevar+eps);
								coeffsq = SQRF(dctblock[2*ey+ex][0][i]);//horizontal
								noisefactor[2*ey+ex][i][1] = coeffsq/(coeffsq+rolloff[i]*noisevar+eps);
								// noisefactor labels are [RGGB subarray][row/col position][0=vert,1=hor]
							}
							}
					//horizontal lines
					if (4*noisevar>(linehvar[0]+linehvar[1])) {//horizontal lines
						for (int i=1; i<8; i++) {
							dctblock[0][0][i] *= 0.5*(noisefactor[0][i][1]+noisefactor[1][i][1]);//or should we use MIN???
							dctblock[1][0][i] *= 0.5*(noisefactor[0][i][1]+noisefactor[1][i][1]);//or should we use MIN???
						}
					}
					if (4*noisevar>(linehvar[2]+linehvar[3])) {//horizontal lines
						for (int i=1; i<8; i++) {
							dctblock[2][0][i] *= 0.5*(noisefactor[2][i][1]+noisefactor[3][i][1]);//or should we use MIN???
							dctblock[3][0][i] *= 0.5*(noisefactor[2][i][1]+noisefactor[3][i][1]);//or should we use MIN???
						}
					}
					//vertical lines
					if (4*noisevar>(linevvar[0]+linevvar[2])) {//vertical lines
						for (int i=1; i<8; i++) {
							dctblock[0][i][0] *= 0.5*(noisefactor[0][i][0]+noisefactor[2][i][0]);//or should we use MIN???
							dctblock[2][i][0] *= 0.5*(noisefactor[0][i][0]+noisefactor[2][i][0]);//or should we use MIN???
						}
					}
					if (4*noisevar>(linevvar[1]+linevvar[3])) {//vertical lines
						for (int i=1; i<8; i++) {
							dctblock[1][i][0] *= 0.5*(noisefactor[1][i][0]+noisefactor[3][i][0]);//or should we use MIN???
							dctblock[3][i][0] *= 0.5*(noisefactor[1][i][0]+noisefactor[3][i][0]);//or should we use MIN???
						}
					}
					for (int ey=0; ey<2; ey++) // (ex,ey) specify RGGB subarray
						for (int ex=0; ex<2; ex++) {
							ddct8x8s(1, dctblock[2*ey+ex]); //inverse DCT
							
							for (int i=0; i<8; i++) 
								for (int j=0; j<8; j++) {
									cfadn[(rr+2*i+ey)*TS+cc+2*j+ex] += window[i]*window[j]*dctblock[2*ey+ex][i][j];
								}
						}	
				}
	}
			// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
			
			// copy smoothed results back to image matrix
			for (int rr=16; rr < rr1-16; rr++)
				for (int cc=16,row=rr+top; cc < cc1-16; cc++) {
					int col = cc + left;
					int indx = row*width + col;
					int c = FC(row,col);
					if (image[indx][c] <65535*0.8*clip_pt && cfadn[(rr)*TS+cc]<0.8*clip_pt)
					image[indx][c] = CLIP((int)(65535*cfadn[(rr)*TS+cc]+ 0.5)); } 
	}
	
	// clean up
	free(buffer);
	}
	// done
	t2 = clock();
	dt = ((double)(t2-t1)) / CLOCKS_PER_SEC;
	if (verbose) {
		fprintf(stderr,_("elapsed time = %5.3fs\n"),dt);
	}
	
	
}
#undef TS
//#undef TA